arrows/mvg/algo/integrate_depth_maps.cxx

```diff
@@ -0,0 +1,336 @@
+#include <arrows/core/depth_utils.h>
+
+#include <vital/util/transform_image.h>
+
+#include <sstream>
+
+using namespace kwiver::vital;
+
+namespace kwiver {
+namespace arrows {
+namespace mvg {
+
+/// Private implementation class
+class integrate_depth_maps::priv
+{
+public:
+  // Constructor
+  priv()
+    : ray_potential_rho(1.0),
+      ray_potential_thickness(20.0),
+      ray_potential_eta(1.0),
+      ray_potential_epsilon(0.01),
+      ray_potential_delta(200.0),
+      grid_spacing {1.0, 1.0, 1.0},
+      voxel_spacing_factor(1.0),
+      m_logger(vital::get_logger("arrows.mvg.integrate_depth_maps"))
+  {
+  }
+
+  // integrate a depth image into the integration volume
+  void integrate_depth_map(image_of<double>& volume,
+                           camera_perspective const& camera,
+                           image_of<double> const& depth,
+                           image_of<double> const& weight,
+                           vector_3d const& origin,
+                           vector_3d const& spacing) const;
+
+  // compute the TSDF ray potential given an estimated depth and real depth
+  double ray_potential(double est_depth, double real_depth) const;
+
+  double ray_potential_rho;
+  double ray_potential_thickness;
+  double ray_potential_eta;
+  double ray_potential_epsilon;
+  double ray_potential_delta;
+
+  int grid_dims[3];
+
+  // Actual spacing is computed as
+  //   voxel_scale_factor * pixel_to_world_scale * grid_spacing
+  // relative spacings per dimension
+  double grid_spacing[3];
+
+  // multiplier on all dimensions of grid spacing
+  double voxel_spacing_factor;
+
+  double const_thickness;
+  double const_delta;
+  double const_slope;
+  double const_freespace_val;
+  double const_occluded_val;
+
+  // Logger handle
+  vital::logger_handle_t m_logger;
+};
+
+// ----------------------------------------------------------------------------
+
+// integrate a depth image into the integration volume
+void
+integrate_depth_maps::priv
+::integrate_depth_map(image_of<double>& volume,
+                      camera_perspective const& camera,
+                      image_of<double> const& depth,
+                      image_of<double> const& weight,
+                      vector_3d const& origin,
+                      vector_3d const& spacing) const
+{
+  auto const ni = volume.width();
+  auto const nj = volume.height();
+  auto const nk = static_cast<int64_t>(volume.depth());
+
+  auto const w = depth.width();
+  auto const h = depth.height();
+
+  #pragma omp parallel for
+  for (int64_t k = 0; k < nk; ++k)
+  {
+    double const z = origin[2] + (k + 0.5) * spacing[2];
+    for (size_t j = 0; j < nj; ++j)
+    {
+      double const y = origin[1] + (j + 0.5) * spacing[1];
+      for (size_t i = 0; i < ni; ++i)
+      {
+        double const x = origin[0] + (i + 0.5) * spacing[0];
+        vector_3d world_pt = vector_3d{ x, y, z };
+        vector_2d image_pt = camera.project(world_pt);
```

Comment:
This could be optimized in the future to break apart the common parts of the projection and reuse some computation across voxels and also avoid duplication of computation with `depth()` in L123 below.  But we can optimize in a follow-up branch.

Reply:
Speeding things up a bit one could increment x (y and z) by the respective amount, e.g. x+=dx, to avoid multiplication and array access costs, but I suspect this is not the slowest part in the loop. I would also avoid the extra copy construction of vector_3d, but define and use a regular constructor, e.g. vector_3d world_pt(x, y, z).

Reply:
I have several changes locally to optimize this.  I've got it running about 10x faster in the special case where there is no lens distortion.  I think it make sense to get this version working and merged before addressing optimizations.